temporal-sdk/src/main/java/io/temporal/common/converter/SearchAttributesUtil.java

```diff
@@ -0,0 +1,187 @@
+    return SearchAttributes.newBuilder()
+        .putAllIndexedFields(objectToPayloadMap(searchAttributes))
+        .build();
+  }
+
+  public static Map<String, Object> decode(SearchAttributes searchAttributes) {
+    if (searchAttributes == null || searchAttributes.getIndexedFieldsCount() == 0) {
+      return null;
+    }
+    return payloadToObjectMap(searchAttributes.getIndexedFieldsMap());
+  }
+
+  public static Map<String, Payload> objectToPayloadMap(Map<String, Object> map) {
+    if (map == null || map.isEmpty()) {
+      return null;
+    }
+    Map<String, Payload> mapOfByteBuffer = new HashMap<>();
+    map.forEach((key, value) -> mapOfByteBuffer.put(key, converter.toData(value).get()));
+    return mapOfByteBuffer;
+  }
+
+  public static Map<String, Object> payloadToObjectMap(Map<String, Payload> serializedMap) {
+    if (serializedMap == null || serializedMap.isEmpty()) {
+      return null;
+    }
+    Map<String, Object> deserializedMap = new HashMap<>();
+    for (Map.Entry<String, Payload> attribute : serializedMap.entrySet()) {
+      String key = attribute.getKey();
+      if (!parseDefaultSearchAttributes(key, attribute.getValue(), deserializedMap)) {
+        String type =
+            attribute
+                .getValue()
+                .getMetadataMap()
+                .get(EncodingKeys.METADATA_TYPE_KEY)
+                .toStringUtf8();
+        Type javaType = encodedTypeToJavaType(type);
+        if (javaType == null) {
+          log.error("Error parsing Search Attribute: {} of type {}.", key, type);
+        } else {
+          deserializedMap.put(
+              key, converter.fromData(attribute.getValue(), javaType.getClass(), javaType));
+        }
+      }
+    }
+    return deserializedMap;
+  }
+
+  public static boolean parseDefaultSearchAttributes(
+      String key, Payload value, Map<String, Object> deserializedMap) {
+    DefaultSearchAttributes searchAttribute;
+    try {
+      searchAttribute = DefaultSearchAttributes.valueOf(key);
+    } catch (IllegalArgumentException e) {
+      return false;
+    }
+    switch (searchAttribute) {
+      case BatcherNamespace:
+      case BatcherUser:
+      case BinaryChecksums:
+      case ExecutionStatus:
+      case RunId:
+      case TaskQueue:
+      case TemporalChangeVersion:
+      case WorkflowId:
+      case WorkflowType:
+        deserializedMap.put(key, converter.fromData(value, String.class, String.class));
+        break;
+      case CloseTime:
+      case ExecutionTime:
+      case StartTime:
+        deserializedMap.put(
+            key, converter.fromData(value, LocalDateTime.class, LocalDateTime.class));
+        break;
+      case ExecutionDuration:
+      case HistoryLength:
+      case StateTransitionCount:
+        deserializedMap.put(key, converter.fromData(value, Integer.class, Integer.class));
+        break;
+    }
+    return true;
+  }
+
+  public static Type encodedTypeToJavaType(String type) {
```

Comment:
encodedTypeToJavaType and javaTypeToEncodedType should be a part of the converter and be private (until we don't need it somewhere else - we can move it our later). No other code should/have to know which server values do we use for java types.